rwengine/src/ai/CharacterController.cpp

```diff
@@ -153,6 +156,222 @@ bool Activities::GoTo::update(CharacterObject *character,
+
+        // We don't want roads with lanes <= 0, also ignore the lastTargetNode
+        if ( (*i) == lastTargetNode || maxLanes <= 0) {
+            i = potentialNodes.erase(i);
+        }
+        else {
+            ++i;
+        }
+    }
+
+    // @todo set the real value
+    const float roadWidth = 5.f;
+
+    // That's a dead end, try to turn around
+    if (potentialNodes.size() == 0) {
+        //@todo try to turn around
+    }
+    // Just a normal road
+    if (potentialNodes.size() == 1) {
+        glm::vec3 pos1 = potentialNodes.at(0)->position;
+        glm::vec3 pos2 = lastTargetNode->position;
+
+        glm::vec3 up = glm::vec3(0.f, 0.f, 1.f);
+        glm::vec3 dir = glm::normalize(pos2 - pos1);
+
+        // Calculate the strafe vector
+        glm::vec3 strafe = glm::cross(up, dir);
+
+        glm::vec3 laneOffset = strafe * ( roadWidth / 2 + roadWidth * (float)(controller->getLane() - 1) );
+
+        realTargetNode = *targetNode;
+        realTargetNode.position += laneOffset;
+    }
+    // Intersection, choose a direction
+    else if (potentialNodes.size() > 1) {
+        // Choose the next node randomly
+        if(nextTargetNode == nullptr) {
+            auto& random = character->engine->randomEngine;
+            int i = std::uniform_int_distribution<>(0, potentialNodes.size() - 1)(random);
+            nextTargetNode = potentialNodes.at(i);
+        }
+
+        controller->nextTargetNode = nextTargetNode;
+
+        glm::vec3 pos1 = nextTargetNode->position;
+        glm::vec3 pos2 = lastTargetNode->position;
+
+        glm::vec3 up = glm::vec3(0.f, 0.f, 1.f);
+        glm::vec3 dir = glm::normalize(pos2 - pos1);
+
+        // Calculate the strafe vector
+        glm::vec3 strafe = glm::cross(up, dir);
+
+        glm::vec3 laneOffset = strafe * ( 2.5f + 5.f * (float)(controller->getLane() - 1) );
+
+        realTargetNode = *targetNode;
+        realTargetNode.position += laneOffset;
+    }
+    // Otherwise set the target to the current node
+    else {
+        realTargetNode = *targetNode;
+    }
+
+    // Check whether we reached the node
+    auto targetDistance =
+        glm::vec2(vehicle->getPosition() - realTargetNode.position);
+
+    const float reachDistance = 5.0f;
+
+    if (glm::length(targetDistance) <= reachDistance) {
+        // Finish the activity
+        return true;
+    } 
+
+    // @todo set real values
+    const float maximumSpeed = 10.f;
+    const float intersectionSpeed = 3.5f;
+
+    float currentSpeed = 0.f;
+
+    // Set the speed depending on where we are driving
+    if ( potentialNodes.size() == 1 ) {
+        currentSpeed = maximumSpeed;
+    }
+    else {
+        currentSpeed = intersectionSpeed;
+    }
+
+    // The minimal distance we test for objects
+    float minColDist = 20.f;
+
+    // Try to stop before pedestrians
+    for (auto& obj : character->engine->pedestrianPool.objects) {
```

Comment:
This is quite an expensive way of checking for an obstruction. This could use the collision detection system to detect pedestrians/vehicles.

Reply:
Yes, I see what you mean, do you think of `WorldCollisionDispatcher`?
But in that case the goal is to prevent a collision, which means I have to sense beforehand if a pedestrian or a vehicle is in front of me, right?

Reply:
Maybe better bullet's raytest/raycast?
See:
http://www.bulletphysics.org/mediawiki-1.5.8/index.php/Using_RayTest

Reply:
Ray testing is one option. Another is a contact test, using ``btCollisionWorld::contactTest`` to check a rectangular area ahead of the vehicle. This would need a bit of work to add a ``GameWorld`` method for doing the actual check, to keep the amount of Bullet that is exposed to objects as low as possible.

We don't know the exact behaviour of the original game (or I don't, at least) so I'm not dead set on a particular solution. Depending on the performance this might be acceptable, with a follow-up issue to make a note that it should be sped up ASAP.

Reply:
btw it can be const.

Reply:
@danhedron When merged I will of course create an issue containing all future tasks.

Reply:
@danhedron, do you think improvements of algorithm for checking should be included in this pull?
 (At least on my laptop (i5 2520m hd3000) fps drop is quite big)

Reply:
I didn't know the performance impact is that great, but merging without collision avoidance would mean accidents on every intersection!
It is actually rather easy to implement a raytest, as there is already the class `VehicleRaycaster` in `VehicleObject`. The problem however is, that the function returns a `btRigidBody` and no `GameObject`, which I would need for lane changing.
Unfortunately I won't get any work done this weekend.

Reply:
Sorry, problem is also on master. I'll make issue, earlier I's getting 60fps on release, now 20-30. :(

Reply:
Nevermind, but I will still try to implement raytest, it should be done right at the first try :+1:

Reply:
I think it's better to put it off. And do it in next pull (when this will be merged).

Reply:
Then it's ready to merge.